code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/PushTemplateImageUtil.kt

```diff
@@ -42,105 +43,158 @@ internal object PushTemplateImageUtil {
+     * @return [Int] number of images that were found in cache or successfully downloaded
+     */
+    internal fun downloadImage(
+        cacheService: CacheService,
+        urlList: List<String?>
+    ): Int {
+        val assetCacheLocation = getAssetCacheLocation()
+        if (urlList.isEmpty() || assetCacheLocation.isNullOrEmpty()) {
+            return 0
+        }
+
+        var downloadedImageCount = 0
+        val latch = CountDownLatch(urlList.size)
+        for (url in urlList) {
+            if (url == null || !UrlUtils.isValidUrl(url)) {
+                latch.countDown()
+                continue
+            }
+
+            val cacheResult = cacheService[assetCacheLocation, url]
+            if (cacheResult != null) {
+                Log.trace(
+                    PushTemplateConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Found cached image for $url"
+                )
+                downloadedImageCount++
+                latch.countDown()
+                continue
+            }
+
+            download(url) { image ->
+                image?.let {
                     Log.trace(
                         PushTemplateConstants.LOG_TAG,
                         SELF_TAG,
-                        "Downloaded push notification image from url ($url)"
+                        "Successfully download image from $url"
                     )
+                    downloadedImageCount++
+
+                    // scale down the bitmap to 300dp x 200dp as we don't want to use a full
+                    // size image due to memory constraints
+                    val pushImage = scaleBitmap(image)
+                    // write bitmap to cache
+                    try {
+                        bitmapToInputStream(pushImage).use { bitmapInputStream ->
+                            cacheBitmapInputStream(
+                                cacheService,
+                                bitmapInputStream,
+                                url
+                            )
+                        }
+                    } catch (exception: IOException) {
+                        Log.trace(
+                            PushTemplateConstants.LOG_TAG,
+                            SELF_TAG,
+                            "Exception occurred creating an input stream from a bitmap for {$url}: ${exception.localizedMessage}."
+                        )
+                    }
                 }
-            } catch (e: IOException) {
-                Log.warning(
-                    PushTemplateConstants.LOG_TAG,
-                    SELF_TAG,
-                    "Failed to download push notification image from url ($url). Exception: ${e.message}"
-                )
-            } finally {
-                connection?.disconnect()
+                latch.countDown()
             }
-            return bitmap
         }
+        latch.await(DOWNLOAD_TIMEOUT_SECS.toLong(), TimeUnit.SECONDS)
```

Comment:
Can you wrap `latch.await` in try/catch to swallow a possible `InterruptedException`